File: leetcode/binary_tree/is_valid_bst.rb
def is_valid_bst_helper(root, stack=[])
  return if tree.nil?

  is_valid_bst_helper(root.left, stack)

  return false if stack[-1] >= root.val
  stack << root.val

  is_valid_bst_helper(root.right, stack)

  true
end

def is_valid_bst(root)
  stack = [-Float::INFINITY]
  is_valid_bst_helper(root, stack)
end

File: leetcode/binary_tree/isValidBST.js
function helper(root, stack) {
  if (root === null) return true;

  const val = root.val;

  const left = helper(root.left, stack);
  if (val > stack[stack.length - 1]) {
    stack.push(val);
  } else {
    return false;
  }
  const right = helper(root.right, stack);

  return left && right;
}

var isValidBST = function(root) {
  const stack = [-Infinity];

  return helper(root, stack);
};
